File: resources/expected/ros2/isolette/strict/src/isolette_cpp_pkg/src/base_code/thermostat_monitor_temperature_manage_alarm_mat_base_src.cpp
#include "isolette_cpp_pkg/base_headers/thermostat_monitor_temperature_manage_alarm_mat_base_src.hpp"

//========================================================
// Re-running Codegen will overwrite changes to this file
//========================================================

thermostat_monitor_temperature_manage_alarm_mat_base::thermostat_monitor_temperature_manage_alarm_mat_base() : Node("thermostat_monitor_temperature_manage_alarm_mat")
{
    cb_group_ = this->create_callback_group(rclcpp::CallbackGroupType::Reentrant);
    subscription_options_.callback_group = cb_group_;

    // Setting up connections
    thermostat_monitor_temperature_manage_alarm_mat_current_tempWstatus_subscription_ = this->create_subscription<isolette_cpp_pkg_interfaces::msg::TempWstatusimpl>(
        "thermostat_monitor_temperature_manage_alarm_mat_current_tempWstatus",
        1,
        std::bind(&thermostat_monitor_temperature_manage_alarm_mat_base::accept_current_tempWstatus, this, std::placeholders::_1), subscription_options_);

    thermostat_monitor_temperature_manage_alarm_mat_lower_alarm_temp_subscription_ = this->create_subscription<isolette_cpp_pkg_interfaces::msg::Tempimpl>(
        "thermostat_monitor_temperature_manage_alarm_mat_lower_alarm_temp",
        1,
        std::bind(&thermostat_monitor_temperature_manage_alarm_mat_base::accept_lower_alarm_temp, this, std::placeholders::_1), subscription_options_);

    thermostat_monitor_temperature_manage_alarm_mat_upper_alarm_temp_subscription_ = this->create_subscription<isolette_cpp_pkg_interfaces::msg::Tempimpl>(
        "thermostat_monitor_temperature_manage_alarm_mat_upper_alarm_temp",
        1,
        std::bind(&thermostat_monitor_temperature_manage_alarm_mat_base::accept_upper_alarm_temp, this, std::placeholders::_1), subscription_options_);

    thermostat_monitor_temperature_manage_alarm_mat_monitor_mode_subscription_ = this->create_subscription<isolette_cpp_pkg_interfaces::msg::MonitorMode>(
        "thermostat_monitor_temperature_manage_alarm_mat_monitor_mode",
        1,
        std::bind(&thermostat_monitor_temperature_manage_alarm_mat_base::accept_monitor_mode, this, std::placeholders::_1), subscription_options_);

    thermostat_monitor_temperature_manage_alarm_mat_alarm_control_publisher_ = this->create_publisher<isolette_cpp_pkg_interfaces::msg::OnOff>(
        "operator_interface_oip_oit_alarm_control",
        1);

    // timeTriggeredCaller callback timer
    periodTimer_ = this->create_wall_timer(std::chrono::milliseconds(1000),
        std::bind(&thermostat_monitor_temperature_manage_alarm_mat_base::timeTriggeredCaller, this), cb_group_);

    // Used by receiveInputs
    inDataPortTupleVector = {
        {&infrastructureIn_current_tempWstatus, &applicationIn_current_tempWstatus},
        {&infrastructureIn_lower_alarm_temp, &applicationIn_lower_alarm_temp},
        {&infrastructureIn_upper_alarm_temp, &applicationIn_upper_alarm_temp},
        {&infrastructureIn_monitor_mode, &applicationIn_monitor_mode}
    };

    // Used by receiveInputs
    inEventPortTupleVector = {
        {&infrastructureIn_current_tempWstatus, &applicationIn_current_tempWstatus},
        {&infrastructureIn_lower_alarm_temp, &applicationIn_lower_alarm_temp},
        {&infrastructureIn_upper_alarm_temp, &applicationIn_upper_alarm_temp},
        {&infrastructureIn_monitor_mode, &applicationIn_monitor_mode}
    };

    // Used by sendOutputs
    outPortTupleVector = {
        {&applicationOut_alarm_control, &infrastructureOut_alarm_control, &thermostat_monitor_temperature_manage_alarm_mat_base::sendOut_alarm_control}
    };

}

void thermostat_monitor_temperature_manage_alarm_mat_base::init_current_tempWstatus(isolette_cpp_pkg_interfaces::msg::TempWstatusimpl val) {
    enqueue(infrastructureIn_current_tempWstatus, val);
}

void thermostat_monitor_temperature_manage_alarm_mat_base::init_lower_alarm_temp(isolette_cpp_pkg_interfaces::msg::Tempimpl val) {
    enqueue(infrastructureIn_lower_alarm_temp, val);
}

void thermostat_monitor_temperature_manage_alarm_mat_base::init_upper_alarm_temp(isolette_cpp_pkg_interfaces::msg::Tempimpl val) {
    enqueue(infrastructureIn_upper_alarm_temp, val);
}

void thermostat_monitor_temperature_manage_alarm_mat_base::init_monitor_mode(isolette_cpp_pkg_interfaces::msg::MonitorMode val) {
    enqueue(infrastructureIn_monitor_mode, val);
}


//=================================================
//  C o m m u n i c a t i o n
//=================================================

void thermostat_monitor_temperature_manage_alarm_mat_base::accept_current_tempWstatus(isolette_cpp_pkg_interfaces::msg::TempWstatusimpl msg)
{
    enqueue(infrastructureIn_current_tempWstatus, msg);
}

void thermostat_monitor_temperature_manage_alarm_mat_base::accept_lower_alarm_temp(isolette_cpp_pkg_interfaces::msg::Tempimpl msg)
{
    enqueue(infrastructureIn_lower_alarm_temp, msg);
}

void thermostat_monitor_temperature_manage_alarm_mat_base::accept_upper_alarm_temp(isolette_cpp_pkg_interfaces::msg::Tempimpl msg)
{
    enqueue(infrastructureIn_upper_alarm_temp, msg);
}

void thermostat_monitor_temperature_manage_alarm_mat_base::accept_monitor_mode(isolette_cpp_pkg_interfaces::msg::MonitorMode msg)
{
    enqueue(infrastructureIn_monitor_mode, msg);
}

isolette_cpp_pkg_interfaces::msg::TempWstatusimpl thermostat_monitor_temperature_manage_alarm_mat_base::get_current_tempWstatus() {
    MsgType msg = applicationIn_current_tempWstatus.front();
    return std::get<isolette_cpp_pkg_interfaces::msg::TempWstatusimpl>(msg);
}

isolette_cpp_pkg_interfaces::msg::Tempimpl thermostat_monitor_temperature_manage_alarm_mat_base::get_lower_alarm_temp() {
    MsgType msg = applicationIn_lower_alarm_temp.front();
    return std::get<isolette_cpp_pkg_interfaces::msg::Tempimpl>(msg);
}

isolette_cpp_pkg_interfaces::msg::Tempimpl thermostat_monitor_temperature_manage_alarm_mat_base::get_upper_alarm_temp() {
    MsgType msg = applicationIn_upper_alarm_temp.front();
    return std::get<isolette_cpp_pkg_interfaces::msg::Tempimpl>(msg);
}

isolette_cpp_pkg_interfaces::msg::MonitorMode thermostat_monitor_temperature_manage_alarm_mat_base::get_monitor_mode() {
    MsgType msg = applicationIn_monitor_mode.front();
    return std::get<isolette_cpp_pkg_interfaces::msg::MonitorMode>(msg);
}

void thermostat_monitor_temperature_manage_alarm_mat_base::sendOut_alarm_control(MsgType msg)
{
    if (auto typedMsg = std::get_if<isolette_cpp_pkg_interfaces::msg::OnOff>(&msg)) {
        thermostat_monitor_temperature_manage_alarm_mat_alarm_control_publisher_->publish(*typedMsg);
    } else {
        PRINT_ERROR("Sending out wrong type of variable on port alarm_control.\nThis shouldn't be possible.  If you are seeing this message, please notify this tool's current maintainer.");
    }
}

void thermostat_monitor_temperature_manage_alarm_mat_base::put_alarm_control(isolette_cpp_pkg_interfaces::msg::OnOff msg)
{
    enqueue(applicationOut_alarm_control, msg);
}

void thermostat_monitor_temperature_manage_alarm_mat_base::timeTriggeredCaller() {
    receiveInputs();
    timeTriggered();
    sendOutputs();
}

void thermostat_monitor_temperature_manage_alarm_mat_base::receiveInputs() {
    for (std::tuple<std::queue<MsgType>*, std::queue<MsgType>*> port : inDataPortTupleVector) {
        auto infrastructureQueue = std::get<0>(port);
        if (!infrastructureQueue->empty()) {
            auto msg = infrastructureQueue->front();
            enqueue(*std::get<1>(port), msg);
        }
    }
    for (std::tuple<std::queue<MsgType>*, std::queue<MsgType>*> port : inEventPortTupleVector) {
        auto infrastructureQueue = std::get<0>(port);
        if (!infrastructureQueue->empty()) {
            auto msg = infrastructureQueue->front();
            infrastructureQueue->pop();
            enqueue(*std::get<1>(port), msg);
        }
    }
}

void thermostat_monitor_temperature_manage_alarm_mat_base::enqueue(std::queue<MsgType>& queue, MsgType val) {
    if (queue.size() >= 1) {
        queue.pop();
    }
    queue.push(val);
}

void thermostat_monitor_temperature_manage_alarm_mat_base::sendOutputs() {
    for (std::tuple<std::queue<MsgType>*, std::queue<MsgType>*, void (thermostat_monitor_temperature_manage_alarm_mat_base::*)(MsgType)> port : outPortTupleVector) {
        auto applicationQueue = std::get<0>(port);
        if (applicationQueue->size() != 0) {
            auto msg = applicationQueue->front();
            applicationQueue->pop();
            enqueue(*std::get<1>(port), msg);
        }
    }

    for (std::tuple<std::queue<MsgType>*, std::queue<MsgType>*, void (thermostat_monitor_temperature_manage_alarm_mat_base::*)(MsgType)> port : outPortTupleVector) {
        auto infrastructureQueue = std::get<1>(port);
        if (infrastructureQueue->size() != 0) {
            auto msg = infrastructureQueue->front();
            infrastructureQueue->pop();
            (this->*std::get<2>(port))(msg);
        }
    }
}

File: resources/expected/ros2/python-isolette_inverted_topics/strict/src/isolette_py_pkg/isolette_py_pkg/base_code/thermostat_monitor_temperature_manage_alarm_mat_base_src.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from queue import Queue
from typing import Union
import threading
from rclpy.callback_groups import ReentrantCallbackGroup
from isolette_py_pkg_interfaces.msg import TempWstatusimpl
from isolette_py_pkg_interfaces.msg import Tempimpl
from isolette_py_pkg_interfaces.msg import MonitorMode
from isolette_py_pkg_interfaces.msg import OnOff

#========================================================
# Re-running Codegen will overwrite changes to this file
#========================================================

class thermostat_monitor_temperature_manage_alarm_mat_base(Node):
    def __init__(self):
        super().__init__("thermostat_monitor_temperature_manage_alarm_mat")

        self.cb_group_ = ReentrantCallbackGroup()

        MsgType = Union[TempWstatusimpl, Tempimpl, MonitorMode, OnOff]
        self.lock_ = threading.Lock()

        # Setting up connections
        self.thermostat_monitor_temperature_manage_alarm_mat_current_tempWstatus_subscription_ = self.create_subscription(
            TempWstatusimpl,
            "temperature_sensor_cpi_thermostat_current_tempWstatus",
            self.accept_current_tempWstatus,
            1,
            callback_group=self.cb_group_)

        self.thermostat_monitor_temperature_manage_alarm_mat_lower_alarm_temp_subscription_ = self.create_subscription(
            Tempimpl,
            "thermostat_monitor_temperature_manage_monitor_interface_mmit_lower_alarm_temp",
            self.accept_lower_alarm_temp,
            1,
            callback_group=self.cb_group_)

        self.thermostat_monitor_temperature_manage_alarm_mat_upper_alarm_temp_subscription_ = self.create_subscription(
            Tempimpl,
            "thermostat_monitor_temperature_manage_monitor_interface_mmit_upper_alarm_temp",
            self.accept_upper_alarm_temp,
            1,
            callback_group=self.cb_group_)

        self.thermostat_monitor_temperature_manage_alarm_mat_monitor_mode_subscription_ = self.create_subscription(
            MonitorMode,
            "thermostat_monitor_temperature_manage_monitor_mode_mmmt_monitor_mode",
            self.accept_monitor_mode,
            1,
            callback_group=self.cb_group_)

        self.thermostat_monitor_temperature_manage_alarm_mat_alarm_control_publisher_ = self.create_publisher(
            OnOff,
            "thermostat_monitor_temperature_manage_alarm_mat_alarm_control",
            1)

        # timeTriggeredCaller callback timer
        self.periodTimer_ = self.create_timer(1000, self.timeTriggeredCaller, callback_group=self.cb_group_)

    def timeTriggered(self):
        pass

        self.infrastructureIn_current_tempWstatus = Queue()
        self.applicationIn_current_tempWstatus = Queue()
        self.infrastructureIn_lower_alarm_temp = Queue()
        self.applicationIn_lower_alarm_temp = Queue()
        self.infrastructureIn_upper_alarm_temp = Queue()
        self.applicationIn_upper_alarm_temp = Queue()
        self.infrastructureIn_monitor_mode = Queue()
        self.applicationIn_monitor_mode = Queue()

        self.infrastructureOut_alarm_control = Queue()
        self.applicationOut_alarm_control = Queue()

        # Used by receiveInputs
        self.inDataPortTupleVector = [
            [self.infrastructureIn_current_tempWstatus, self.applicationIn_current_tempWstatus],
            [self.infrastructureIn_lower_alarm_temp, self.applicationIn_lower_alarm_temp],
            [self.infrastructureIn_upper_alarm_temp, self.applicationIn_upper_alarm_temp],
            [self.infrastructureIn_monitor_mode, self.applicationIn_monitor_mode]
        ]

        # Used by receiveInputs
        self.inEventPortTupleVector = [
            [self.infrastructureIn_current_tempWstatus, self.applicationIn_current_tempWstatus],
            [self.infrastructureIn_lower_alarm_temp, self.applicationIn_lower_alarm_temp],
            [self.infrastructureIn_upper_alarm_temp, self.applicationIn_upper_alarm_temp],
            [self.infrastructureIn_monitor_mode, self.applicationIn_monitor_mode]
        ]

        # Used by sendOutputs
        self.outPortTupleVector = [
            [self.applicationOut_alarm_control, self.infrastructureOut_alarm_control, self.sendOut_alarm_control]
        ]

    #=================================================
    #  C o m m u n i c a t i o n
    #=================================================

    def accept_current_tempWstatus(self, msg):
        typedMsg = TempWstatusimpl()
        typedMsg.data = msg
        self.enqueue(infrastructureIn_current_tempWstatus, msg)

    def accept_lower_alarm_temp(self, msg):
        typedMsg = Tempimpl()
        typedMsg.data = msg
        self.enqueue(infrastructureIn_lower_alarm_temp, msg)

    def accept_upper_alarm_temp(self, msg):
        typedMsg = Tempimpl()
        typedMsg.data = msg
        self.enqueue(infrastructureIn_upper_alarm_temp, msg)

    def accept_monitor_mode(self, msg):
        typedMsg = MonitorMode()
        typedMsg.data = msg
        self.enqueue(infrastructureIn_monitor_mode, msg)

    def get_current_tempWstatus(self):
        msg = applicationIn_current_tempWstatus.front()
        return get(msg)

    def get_lower_alarm_temp(self):
        msg = applicationIn_lower_alarm_temp.front()
        return get(msg)

    def get_upper_alarm_temp(self):
        msg = applicationIn_upper_alarm_temp.front()
        return get(msg)

    def get_monitor_mode(self):
        msg = applicationIn_monitor_mode.front()
        return get(msg)

    def sendOut_alarm_control(self, msg):
        if type(msg) is OnOff:
            typedMsg = OnOff()
            typedMsg.data = msg
            self.thermostat_monitor_temperature_manage_alarm_mat_alarm_control_publisher_.publish(typedMsg)
        else:
            self.get_logger().error("Sending out wrong type of variable on port alarm_control.\nThis shouldn't be possible.  If you are seeing this message, please notify this tool's current maintainer.")

    def put_alarm_control(self, msg):
        typedMsg = OnOff()
        typedMsg.data = msg
        self.enqueue(self.applicationOut_alarm_control, typedMsg)

    def timeTriggeredCaller(self):
        self.receiveInputs()
        self.timeTriggered()
        self.sendOutputs()

    def receiveInputs(self):
        for port in self.inDataPortTupleVector:
            infrastructureQueue = port[0]
            if not(infrastructureQueue.empty()):
                msg = infrastructureQueue.front()
                self.enqueue(*port[1], msg)

        for port in self.inEventPortTupleVector:
            infrastructureQueue = port[0]
            if not(infrastructureQueue.empty()):
                msg = infrastructureQueue.front()
                infrastructureQueue.pop()
                self.enqueue(port[1], msg)

    def enqueue(self, queue, val):
        if queue.size() >= 1:
            queue.pop()
        queue.push(val)

    def sendOutputs(self):
        for port in self.outPortTupleVector:
            applicationQueue = port[0]
            if applicationQueue.size() != 0:
                msg = applicationQueue.front()
                applicationQueue.pop()
                self.enqueue(port[1], msg)

        for port in self.outPortTupleVector:
            infrastructureQueue = port[1]
            if infrastructureQueue.size() != 0:
                msg = infrastructureQueue.front()
                infrastructureQueue.pop()
                (port[2])(msg)
